crates/fuel-core/src/graphql_api/storage/coins.rs

```diff
@@ -23,6 +41,180 @@ pub fn owner_coin_id_key(owner: &Address, coin_id: &UtxoId) -> OwnedCoinKey {
+pub struct CoinsToSpendIndexKey([u8; CoinsToSpendIndexKey::LEN]);
+
+impl Default for CoinsToSpendIndexKey {
+    fn default() -> Self {
+        Self([0u8; CoinsToSpendIndexKey::LEN])
+    }
+}
+
+impl core::fmt::Display for CoinsToSpendIndexKey {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(
+            f,
+            "owner={}, asset_id={}, retryable_flag={}, amount={}",
+            self.owner(),
+            self.asset_id(),
+            self.retryable_flag(),
+            self.amount()
+        )
+    }
+}
+
+impl CoinsToSpendIndexKey {
+    const LEN: usize = Address::LEN
+        + AssetId::LEN
+        + u8::BITS as usize / 8
+        + u64::BITS as usize / 8
+        + TxId::LEN
+        + 2;
+
+    #[allow(clippy::arithmetic_side_effects)]
+    pub fn from_coin(coin: &Coin) -> Self {
+        let address_bytes = coin.owner.as_ref();
+        let asset_id_bytes = coin.asset_id.as_ref();
+        let amount_bytes = coin.amount.to_be_bytes();
+        let utxo_id_bytes = utxo_id_to_bytes(&coin.utxo_id);
+
+        let mut arr = [0; CoinsToSpendIndexKey::LEN];
+        let mut offset = 0;
+        arr[offset..offset + Address::LEN].copy_from_slice(address_bytes);
+        offset += Address::LEN;
+        arr[offset..offset + AssetId::LEN].copy_from_slice(asset_id_bytes);
+        offset += AssetId::LEN;
+        arr[offset..offset + u8::BITS as usize / 8].copy_from_slice(&NON_RETRYABLE_BYTE);
+        offset += u8::BITS as usize / 8;
+        arr[offset..offset + u64::BITS as usize / 8].copy_from_slice(&amount_bytes);
+        offset += u64::BITS as usize / 8;
+        arr[offset..].copy_from_slice(&utxo_id_bytes);
+        Self(arr)
+    }
+
+    #[allow(clippy::arithmetic_side_effects)]
+    pub fn from_message(message: &Message, base_asset_id: &AssetId) -> Self {
+        let address_bytes = message.recipient().as_ref();
+        let asset_id_bytes = base_asset_id.as_ref();
+        let amount_bytes = message.amount().to_be_bytes();
+        let nonce_bytes = message.nonce().as_slice();
+
+        let mut arr = [0; CoinsToSpendIndexKey::LEN];
+        let mut offset = 0;
+        arr[offset..offset + Address::LEN].copy_from_slice(address_bytes);
+        offset += Address::LEN;
+        arr[offset..offset + AssetId::LEN].copy_from_slice(asset_id_bytes);
+        offset += AssetId::LEN;
+        arr[offset..offset + u8::BITS as usize / 8].copy_from_slice(
+            if message.is_retryable_message() {
+                &RETRYABLE_BYTE
+            } else {
+                &NON_RETRYABLE_BYTE
+            },
+        );
+        offset += u8::BITS as usize / 8;
+        arr[offset..offset + u64::BITS as usize / 8].copy_from_slice(&amount_bytes);
+        offset += u64::BITS as usize / 8;
+        arr[offset..offset + Nonce::LEN].copy_from_slice(nonce_bytes);
+        offset += Nonce::LEN;
+        arr[offset..].copy_from_slice(&indexation::coins_to_spend::MESSAGE_PADDING_BYTES);
```

Comment:
Hmm, how well RocksDB works with the dynamic sized keys? Maybe based on the type of the message/coin we could use 32/34 bytes for `Nonce`/`UtxoId` types and during decoding decide what type to return? 

Just a thought, if it is hard to support or implement, I'm okay with the current padding approach =)

Reply:
Actually, in the complete PR I decided to remove the padding approach and use variable length keys. I don't know about any performance implications on RocksDB side. Also, we never query for a large amounts of data (255 items at most with the current limits), so we should be good. Taking this into consideration I though it's not worth "wasting" additional two bytes for _every_ indexed message.